extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java

```diff
@@ -132,12 +133,12 @@ public QueryDefinition makeQueryDefinition(
       shuffleSpecFactoryPreAggregation = intermediateClusterBy.isEmpty()
                                          ? ShuffleSpecFactories.singlePartition()
                                          : ShuffleSpecFactories.globalSortWithMaxPartitionCount(maxWorkerCount);
-      shuffleSpecFactoryPostAggregation = doLimitOrOffset
+      shuffleSpecFactoryPostAggregation = doLimitOrOffset && !needsFinalShuffling
```

Comment:
Same reason as on the other line - Consider a query like: 

```sql
SELECT col1 FROM foo GROUP BY col1 LIMIT 1
```

If the group by returns 10 partitions, the limit would work on them individually, which is not desirable. I think this change isn't required as we are always setting up limit at the end.

Reply:
Since it is a subquery, the needsFinalShuffling should always be false here, so I believe that case should still work.

Reply:
My example was incorrect, but I am still doubtful. The only reason that this change seems incorrect is what happens if `doLimitOrOffset=true` but the factory doesn't generate a single partition. 
for example: `EXPORT ... SELECT col1 FROM foo GROUP BY col1 LIMIT 1`
In that case, the LIMIT wouldn't be correct

Limit processor must always receive a single partition, and having `ShuffleSpecFactories.singlePartition` ensures that. The earlier code ensured that. The new code doesn't ensure that, and it can at best have the same result as the old code, or be incorrect (as is my guess). Alternatively, in which case will this change be useful over the original code.